File: api/app/plan/geometry.py
from __future__ import annotations

import uuid

from ..schemas import HouseSpec, PlanEdge, PlanGraph, PlanRoom, Rect


def _rid() -> str:
    return str(uuid.uuid4())


def generate_plan_graph(spec: HouseSpec) -> PlanGraph:
    """
    Deterministic MVP layout.

    Coordinate system in feet:
    - origin at top-left
    - x to the right, y down
    """
    outline_w = 52.0
    outline_h = 34.0
    outline = Rect(x=0, y=0, w=outline_w, h=outline_h)

    left_w = 32.0
    right_w = outline_w - left_w

    rooms: list[PlanRoom] = []
    edges: list[PlanEdge] = []
    warnings: list[str] = []

    # Zone A (left): public rooms stacked
    y = 0.0
    public = [r for r in spec.rooms if r.type in {"living", "kitchen", "dining"}]
    if not public:
        warnings.append("No public rooms (living/kitchen/dining) in spec; adding default Great Room.")
        public = []
        public.append(type(spec.rooms[0]).model_validate({"id": _rid(), "type": "living", "name": "Great Room", "area_ft2": 320}))  # type: ignore[attr-defined]

    # Assign heights by area / width; clamp to keep readable rectangles.
    for r in public:
        h = max(8.0, min(14.0, r.area_ft2 / left_w))
        if y + h > outline_h:
            break
        rooms.append(
            PlanRoom(
                id=r.id,
                name=r.name,
                type=r.type,
                area_ft2=r.area_ft2,
                rect_ft=Rect(x=0, y=y, w=left_w, h=h),
            )
        )
        y += h

    # Add a small entry/hall connector if space permits.
    hall_h = max(4.0, outline_h - y)
    hall_id = _rid()
    if hall_h >= 4.0:
        rooms.append(
            PlanRoom(
                id=hall_id,
                name="Hall",
                type="hall",
                area_ft2=left_w * hall_h,
                rect_ft=Rect(x=0, y=y, w=left_w, h=hall_h),
            )
        )

    # Zone B (right): bedrooms + baths stacked
    y2 = 0.0
    priv = [r for r in spec.rooms if r.type in {"bedroom", "bathroom", "laundry"}]
    if not priv:
        warnings.append("No private rooms (bedroom/bathroom/laundry) in spec; adding defaults.")

    for r in priv:
        h = max(6.0, min(12.0, r.area_ft2 / right_w))
        if y2 + h > outline_h:
            break
        rooms.append(
            PlanRoom(
                id=r.id,
                name=r.name,
                type=r.type,
                area_ft2=r.area_ft2,
                rect_ft=Rect(x=left_w, y=y2, w=right_w, h=h),
            )
        )
        y2 += h

    # Edges: naive adjacency based on types.
    living = next((r for r in rooms if r.type == "living"), None)
    kitchen = next((r for r in rooms if r.type == "kitchen"), None)
    dining = next((r for r in rooms if r.type == "dining"), None)
    if living and kitchen:
        edges.append(PlanEdge(a=living.id, b=kitchen.id))
    if kitchen and dining:
        edges.append(PlanEdge(a=kitchen.id, b=dining.id))
    if living and dining:
        edges.append(PlanEdge(a=living.id, b=dining.id))

    # Hall connects to first bedroom if present
    first_bed = next((r for r in rooms if r.type == "bedroom"), None)
    if first_bed and hall_id:
        edges.append(PlanEdge(a=hall_id, b=first_bed.id, kind="circulation"))

    return PlanGraph(outline_ft=outline, rooms=rooms, edges=edges, warnings=warnings)

File: functions/src/geometry.ts
import { randomUUID } from "node:crypto";

import type { HouseSpec, PlanEdge, PlanGraph, PlanRoom, Rect } from "./types";

export function generatePlanGraph(spec: HouseSpec): PlanGraph {
  const outlineW = 52.0;
  const outlineH = 34.0;
  const outline: Rect = { x: 0, y: 0, w: outlineW, h: outlineH };

  const leftW = 32.0;
  const rightW = outlineW - leftW;

  const rooms: PlanRoom[] = [];
  const edges: PlanEdge[] = [];
  const warnings: string[] = [];

  let y = 0.0;
  let publicRooms = spec.rooms.filter((r) => ["living", "kitchen", "dining"].includes(r.type));
  if (publicRooms.length === 0) {
    warnings.push("No public rooms in spec; added a default Great Room.");
    publicRooms = [
      {
        id: randomUUID(),
        type: "living",
        name: "Great Room",
        area_ft2: 320,
      },
    ];
  }

  for (const r of publicRooms) {
    const h = clamp(r.area_ft2 / leftW, 8.0, 14.0);
    if (y + h > outlineH) break;
    rooms.push({
      id: r.id,
      name: r.name,
      type: r.type,
      area_ft2: r.area_ft2,
      rect_ft: { x: 0, y, w: leftW, h },
    });
    y += h;
  }

  const hallH = Math.max(4.0, outlineH - y);
  const hallId = randomUUID();
  if (hallH >= 4.0) {
    rooms.push({
      id: hallId,
      name: "Hall",
      type: "hall",
      area_ft2: leftW * hallH,
      rect_ft: { x: 0, y, w: leftW, h: hallH },
    });
  }

  let y2 = 0.0;
  const privateRooms = spec.rooms.filter((r) =>
    ["bedroom", "bathroom", "laundry"].includes(r.type),
  );
  if (privateRooms.length === 0) {
    warnings.push("No private rooms in spec; private zone is empty.");
  }
  for (const r of privateRooms) {
    const h = clamp(r.area_ft2 / rightW, 6.0, 12.0);
    if (y2 + h > outlineH) break;
    rooms.push({
      id: r.id,
      name: r.name,
      type: r.type,
      area_ft2: r.area_ft2,
      rect_ft: { x: leftW, y: y2, w: rightW, h },
    });
    y2 += h;
  }

  const living = rooms.find((r) => r.type === "living");
  const kitchen = rooms.find((r) => r.type === "kitchen");
  const dining = rooms.find((r) => r.type === "dining");
  if (living && kitchen) edges.push({ a: living.id, b: kitchen.id, kind: "adjacent" });
  if (kitchen && dining) edges.push({ a: kitchen.id, b: dining.id, kind: "adjacent" });
  if (living && dining) edges.push({ a: living.id, b: dining.id, kind: "adjacent" });

  const firstBed = rooms.find((r) => r.type === "bedroom");
  if (firstBed) edges.push({ a: hallId, b: firstBed.id, kind: "circulation" });

  return {
    version: "1.0",
    outline_ft: outline,
    rooms,
    edges,
    warnings,
  };
}

export function renderPlanSvg(plan: PlanGraph, pxPerFt = 12): string {
  const w = Math.round(plan.outline_ft.w * pxPerFt);
  const h = Math.round(plan.outline_ft.h * pxPerFt);
  const parts: string[] = [];

  parts.push(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${w}" height="${h}" viewBox="0 0 ${w} ${h}">`,
  );
  parts.push('<rect width="100%" height="100%" fill="#f8fafc"/>');
  parts.push(
    `<rect x="8" y="8" width="${w - 16}" height="${h - 16}" fill="none" stroke="#0f172a" stroke-width="3"/>`,
  );

  for (const room of plan.rooms) {
    const x = 8 + room.rect_ft.x * pxPerFt;
    const y = 8 + room.rect_ft.y * pxPerFt;
    const rw = room.rect_ft.w * pxPerFt;
    const rh = room.rect_ft.h * pxPerFt;
    const label = escapeXml(room.name);
    parts.push(
      `<rect x="${x.toFixed(1)}" y="${y.toFixed(1)}" width="${rw.toFixed(1)}" height="${rh.toFixed(1)}" fill="white" stroke="#0f172a" stroke-width="2"/>`,
    );
    parts.push(
      `<text x="${(x + rw / 2).toFixed(1)}" y="${(y + rh / 2).toFixed(1)}" text-anchor="middle" dominant-baseline="middle" font-family="ui-sans-serif,system-ui" font-size="14" fill="#0f172a">${label}</text>`,
    );
  }

  parts.push("</svg>");
  return parts.join("\n");
}

function clamp(n: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, n));
}

function escapeXml(raw: string): string {
  return raw
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;");
}
